src/Swarm/Game/Achievement/Persistence.hs

```diff
@@ -30,23 +28,25 @@ getSwarmAchievementsPath createDirs = getSwarmXdgDataSubdir createDirs "achievem
 -- | Load saved info about achievements from XDG data directory.
 -- Returns a tuple of warnings and attained achievements.
 loadAchievementsInfo ::
-  IO ([SystemFailure], [Attainment])
+  (Has (Accum (Seq SystemFailure)) sig m, Has (Lift IO) sig m) =>
+  m [Attainment]
 loadAchievementsInfo = do
-  savedAchievementsPath <- getSwarmAchievementsPath False
-  doesParentExist <- doesDirectoryExist savedAchievementsPath
+  savedAchievementsPath <- sendIO $ getSwarmAchievementsPath False
+  doesParentExist <- sendIO $ doesDirectoryExist savedAchievementsPath
   if doesParentExist
     then do
-      contents <- listDirectory savedAchievementsPath
-      eithersList <- forM contents $ \p -> do
+      contents <- sendIO $ listDirectory savedAchievementsPath
+      forMW contents $ \p -> do
         let fullPath = savedAchievementsPath </> p
-        isFile <- doesFileExist fullPath
+        isFile <- sendIO $ doesFileExist fullPath
         if isFile
           then do
             eitherDecodedFile <- sendIO (Y.decodeFileEither fullPath)
             return $ left (AssetNotLoaded Achievement p . CanNotParse) eitherDecodedFile
           else return . Left $ AssetNotLoaded Achievement p (EntryNot File)
-      return $ partitionEithers eithersList
-    else return ([AssetNotLoaded Achievement "." $ DoesNotExist Directory], [])
+    else do
+      warn $ AssetNotLoaded Achievement "." $ DoesNotExist Directory
```

Comment:
I like having some kind of writer monad to propagate warnings.  Passing them all explicitly as function return values was pretty clunky.

Reply:
Indeed, that was one of my major motivations for doing this refactoring.